### PR TITLE
Proposal: middleware-oriented design

### PR DESCRIPTION
**Pull request by dougmoscrop**
I have not written tests yet, as the goal is to discuss design but with working code.

```js
const Handler = require('lambda-handler-as-promised');

module.exports.handler = Handler()
  .use(Handler.sanitizerErrors)
  .use(Handler.logging) // these should probably be constructors e.g. .logging()
  .use(function(event, context, next) {
    return next(); // no-op
  })
  .use(co.wrap(function* ifYouWant(event, context, next) {
    try {
      yield next(); // yield* does not work..
    } catch (e) {
       // re throw, return Promise, etc.
    }
  }))
  .use(function() {
    return 'this will be the lambda result'; // of course can do new Promise, etc.
  })
```

### CODE REVIEW

**Comment by AntonBazhal** [comment]
That's cool. Some thoughts:
1. We can, probably, do `co.wrap` ourselves, so generators are supported out of the box
1. I don't really like the way error handlers need to be written. It's somewhat similar to koa, I guess, but still not too intuitive.
1. I'd prefer not to have any context extensions in the main repo

**Comment by dougmoscrop** [comment]
1. I am on the fence, as if nobody is using it (they just use promises) or they use async/await, its bundled for no reason.
2. Hmm, it is the same as koa, for generators, because that's how you have control flow; I think you can also do return next().catch(...) not 100% sure yet. What would you want instead?
3. Yep - just didn't want to get carried away or to put a bunch of your code in my name..

**Comment by AntonBazhal** [comment]
1. Fair enough
1. I wish I knew. Let me think tomorrow.
1. 👍

**Comment by AntonBazhal** [comment]
Just as an idea. What about having two separate lists (`middleware` and `errMiddleware`)? if middleware has signature `(event, context, next)`, it goes to `middleware`, if `(err, event, context, next)` - to `errMiddleware`. So, it's closer to what `express` does.

**Comment by dougmoscrop** [comment]
Yeah, I thought about having two different middleware too. Superficially, I avoided it because it's what Express does, and "people" moved on from Express to Koa and you can still do this:

```js
app.use(function(event, context, next) {
  return next().catch(e => console.log('but I am not error middleware!'))
})
```
I thought in express, this system caused complexity around the order of which you registered handlers and error handlers. Maybe for lambda, we don't care because the order will always be simple? 

It seems to me that we can support errorMiddleware, and people just don't have to use it -- which is OK. It's just a separate chain that runs if dispatch fails.

**Comment by dougmoscrop** [comment]
For fun, here's the version that mostly supports errorMiddleware (not sure of corner cases, need to flush out in tests):

(Note: for basic demo, I put `err` at end of function signature instead of beginning)

```js
'use strict';

// modified from github.com/koajs/compose
function compose(middleware, current) {
	let index = -1;
	function dispatch (i) {
		try {
			const fn = middleware[i] || Function.prototype;
			const next = (err, context) => {
				if (err) {
					return Promise.reject(err);
				}
				if (context) {
					Object.assign(current, { context, err: null });
				}
				return dispatch(i + 1);
			};

			if (i <= index) {
				throw new Error('next() called more than once');
			}
			index = i;

			return Promise.resolve(fn(current.event, current.context, next, current.err))
				.then(result => {
					if (result !== undefined) {
						Object.assign(current, { result });
					}
					return current.result;
				});
		} catch (err) {
			return Promise.reject(err);
		}
	}

	return dispatch(0);
}

module.exports = function () {
	const middleware = [];
	const errorMiddleware = [];

	const handler = (event, context, callback) => {
		const current = { event, context };

		compose(middleware, current)
			.catch(err => compose(errorMiddleware, Object.assign(current, { result: null, err })))
			.then(() => callback(null, current.result))
			.catch(err => callback(err));
	};

	handler.use = mw => {
		const fn = mw.__generatorFunction__ ? mw.__generatorFunction__ : mw;

		if (fn.length === 4) {
			errorMiddleware.push(mw);
		} else {
			middleware.push(mw);
		}

		return handler;
	};

	return handler;
};
```

I'm not super crazy about the complexity.

**Comment by AntonBazhal** [comment]
Looks like with this implementation if you call `next(err)` from error handler, handler would exit with error instead of calling next error handler.

**Comment by dougmoscrop** [comment]
Yeah, I am having a hard time ensuring the control flow in such cases works everywhere - fix one spot, it breaks in another. Since you can do (similar to above):

```js
.use(function(event, context, next) {
  return next().catch(e => {
    throw e;
  });
})
.use(function(err, event, context, next) {
  throw err;
})
.use(function(err, event, context, next) {
  // expect to be called last?
})
.use(function(() => Promise.reject('testing'))
```

**Comment by dougmoscrop** [comment]
I pushed changes that are sort of the minimum - in other words, because you can write an error handler both ways, I am omitting the 'separate way' for now, just because I'd like to make progress.

I also removed the whole "next called more than once" checking, since that could actually be a thing:

next().catch(e => next()) // retry on some kind of failure?

File: lib/handler.js [hunk 1]
```diff
@@ -1,56 +1,42 @@
...
-		const newContext = loggerExtension(context);
-
-		return Promise
-			.resolve(new Promise(
-				(resolve, reject) => {
-					if (options.onBefore) {
-						options.onBefore(event, newContext);
-					}
+	const handler = (event, context, callback) => {
+		dispatch({ event, context })
```
Comment (AntonBazhal) [inline lib/handler.js]: I'd add `return`

File: lib/handler.js [hunk 2]
```diff
@@ -1,56 +1,42 @@
...
-				.then(result => callback(null, result))
-				.catch(err => {
-					return Promise
-						.resolve(utils.callErrorHook(options.onError, err, event, newContext))
-						.then(result => utils.callHook(options.onAfter, result, event, newContext))
-						.then(result => callback(null, result));
-				})
-			)
-			.catch(err => callback(utils.sanitizeError(err, options)));
+	handler.use = mw => {
```
Comment (AntonBazhal) [inline lib/handler.js]: Should it have some checks to be sure that mw is a function?

File: lib/handler.js [hunk 3]
```diff
@@ -1,56 +1,42 @@
...
-		}
-	});
-
-	if (options.errorStack && typeof options.errorStack !== 'boolean') {
-		throw new Error('options.errorStack must be a boolean when present');
+module.exports = function () {
+	const middleware = [];
+
+	// modified from github.com/koajs/compose
+	function dispatch (state, i = 0) {
```
Comment (AntonBazhal) [inline lib/handler.js]: I don't think that this works with Node < 6. Do we care?

**Comment by AntonBazhal** [comment]
Agree. Let's move on.

File: lib/handler.js (hunk 3 above)
Comment (dougmoscrop) [reply lib/handler.js]: I don't think so, but it's so trivial to change, I just made it passed in explicitly.

**Comment by dougmoscrop** [comment]
I did not update documentation.

File: .eslintrc [hunk 4]
```diff
@@ -11,9 +11,6 @@ extends:
 parserOptions:
     sourceType: "scripts"
 
-ecmaFeatures:
```
Comment (dougmoscrop) [inline .eslintrc:14]: eslint @4.x warns about this

File: test/lambda-handler-tests.js [hunk 5]
```diff
@@ -13,44 +15,40 @@ const testContext = {
...
-			expect(wrappedHandler).to.have.length(3);
+	describe('use()', function() {
+		it('should throw when middleware is not provided', function() {
+			expect(() => lambdaHandler().use()).to.throw(Error, /^middleware is not a function$/);
 		});
 
-		it('wrapped handler should return promise', function() {
-			const fixture = lambdaHandler(() => true);
-			expect(fixture({}, testContext, () => {})).to.be.an.instanceof(Promise);
+		it('should throw when midleware is not a non-function', function() {
```
Comment (AntonBazhal) [inline test/lambda-handler-tests.js]: wording problem

File: test/lambda-handler-tests.js [hunk 6]
```diff
@@ -13,44 +15,40 @@ const testContext = {
...
-			fixture({}, testContext, (err, result) => {
-				try {
-					expect(err).to.not.exist;
-					expect(result).to.equal(testResult);
-					done();
-				} catch (assertErr) {
-					done(assertErr);
-				}
-			});
+			expect(result).to.equal(handler);
```
Comment (AntonBazhal) [inline test/lambda-handler-tests.js]: `deep.equal`?

File: test/lambda-handler-tests.js [hunk 7]
```diff
@@ -110,410 +108,126 @@ describe('lambda-handler-as-promised', function() {
...
-				});
-			});
 		});
 
-		describe('errorStack', function() {
-			it('should throw when "errorStack" is not boolean', function() {
-				expect(() => lambdaHandler(() => {}, { errorStack: 'yes' }))
-					.to.throw(Error, /^options.errorStack must be a boolean when present$/);
+		it('lets middleware catch errors', function() {
+			const mock = sinon.mock();
```
Comment (AntonBazhal) [inline test/lambda-handler-tests.js]: Feels messy to have spy, stub and mock in one test. Especially when mock does not check anything. Mocks are useful sometimes, but my experience with them is not that great. According to the docs, mocks should only be used when you need both: validate input and stub behaviour. Otherwise, just use spies or stubs.

File: test/lambda-handler-tests.js [hunk 8]
```diff
@@ -13,44 +15,40 @@ const testContext = {
...
-		});
+	it('should return a function', function() {
+		expect(lambdaHandler()).to.be.a('function');
+	});
 
-		it('should throw when handler is not a function', function() {
-			expect(() => lambdaHandler({})).to.throw(Error, /^handler is not a function$/);
-		});
+	it('should have a use() method', function() {
+		expect(lambdaHandler().use).to.be.a('function');
```
Comment (AntonBazhal) [inline test/lambda-handler-tests.js]: Just preference
```js
expect(lambdaHandler())
  .to.have.property('use')
  .that.is.a('function');
```

File: .eslintrc (hunk 4 above)
Comment (AntonBazhal) [reply .eslintrc:14]: 👍

File: test/lambda-handler-tests.js [hunk 9]
```diff
@@ -110,410 +110,129 @@ describe('lambda-handler-as-promised', function() {
...
-							expect(err).to.be.equal(null);
-							expect(result).to.deep.equal(testError);
-							done();
-						} catch (assertErr) {
-							done(assertErr);
-						}
-					});
+			return fixture(testEvent, testContext, noop)
+				.then(() => {
+					expect(stub)
```
Comment (AntonBazhal) [inline test/lambda-handler-tests.js]: I'm not so crazy) When you assert stubs, just
```js
expect(stub.called);
// or
expect(stub.called).to.be.true;
// or
sinon.assert.called(stub);
```

**Comment by AntonBazhal** [comment]
Just minor thing with stub check

**Comment by dougmoscrop** [comment]
done

**Review by AntonBazhal — APPROVED** [review]
What about docs?

**Comment by dougmoscrop** [comment]
will update

File: README.md [hunk 10]
```diff
@@ -39,180 +39,20 @@ With `lambda-handler-as-promised` you should not worry about top-level error han
...
-}
-
-const handler = lambdaHandler(function(event, context) {
-	context.log.info('This is base context');
-	doSomething();
-	context.log.info('This is base context again');
-	return true;
-});
+module.exports.handler = lambdaHandler()
+	.use(function(evenet, context, next) {
```
Comment (AntonBazhal) [inline README.md:43]: typo in event

File: README.md [hunk 11]
```diff
@@ -6,7 +6,7 @@
 [![Dependencies Status][dependencies-image]][dependencies-url]
 [![DevDependencies Status][devdependencies-image]][devdependencies-url]
 
-Tiny wrapper that ensures that [AWS Lambda][aws-lambda-url] function's callback is always called. In other words, from your handler you can return value, promise, throw exception, and this library will wrap your code into a promise while calling appropriate lambda-required callback. If you’d like, you can call callback by yourself, `lambda-handler-as-promised` will still behave in the same way.
+Tiny wrapper that ensures that [AWS Lambda][aws-lambda-url] function's callback is always called. In other words, from your handler you can return value, promise, throw exception, and this library will wrap your code into a promise while calling appropriate lambda-required callback. Your handler is composed of middleware, similar to Express or Koa.
```
Comment (AntonBazhal) [inline README.md:9]: I'd add links to Express and Koa

File: README.md [hunk 12]
```diff
@@ -39,180 +39,20 @@ With `lambda-handler-as-promised` you should not worry about top-level error han
...
-	context.log.info('This is child context with newField');
-}
-
-const handler = lambdaHandler(function(event, context) {
-	context.log.info('This is base context');
-	doSomething();
-	context.log.info('This is base context again');
-	return true;
-});
+module.exports.handler = lambdaHandler()
```
Comment (AntonBazhal) [inline README.md:42]: Can you, please, add an example of custom error handling?

**Comment by AntonBazhal** [comment]
Probably, it's not a bad idea to add section with "official" middleware. First candidate [https://www.npmjs.com/package/lambda-handler-as-promised-bunyan](https://www.npmjs.com/package/lambda-handler-as-promised-bunyan)

Link does not work, since it's not published yet.

**Comment by dougmoscrop** [comment]
Any idea for a shorter name before going all in on `lambda-handler-as-promised-xyz` ?

**Comment by AntonBazhal** [comment]
Do you mean for this repo? Thought about that, but changing npm name requires publishing of a new library. Ideas?
